transports/webrtc/src/tokio/transport.rs

```diff
@@ -0,0 +1,601 @@
+    ///
+    /// `None` if the socket is only listening on a single interface.
+    if_watcher: IfWatcher,
+}
+
+impl ListenStream {
+    /// Constructs a `WebRTCListenStream` for incoming connections.
+    fn new(
+        listener_id: ListenerId,
+        config: Config,
+        udp_mux: UDPMuxNewAddr,
+        id_keys: identity::Keypair,
+        if_watcher: IfWatcher,
+    ) -> Self {
+        ListenStream {
+            listener_id,
+            listen_addr: udp_mux.listen_addr(),
+            config,
+            udp_mux,
+            id_keys,
+            report_closed: None,
+            if_watcher,
+        }
+    }
+
+    /// Report the listener as closed in a [`TransportEvent::ListenerClosed`] and
+    /// terminate the stream.
+    fn close(&mut self, reason: Result<(), Error>) {
+        match self.report_closed {
+            Some(_) => log::debug!("Listener was already closed."),
+            None => {
+                // Report the listener event as closed.
+                let _ = self
+                    .report_closed
+                    .insert(Some(TransportEvent::ListenerClosed {
+                        listener_id: self.listener_id,
+                        reason,
+                    }));
+            }
+        }
+    }
+
+    fn poll_if_watcher(&mut self, cx: &mut Context<'_>) -> Poll<<Self as Stream>::Item> {
+        while let Poll::Ready(event) = self.if_watcher.poll_if_event(cx) {
+            match event {
+                Ok(IfEvent::Up(inet)) => {
+                    let ip = inet.addr();
+                    if self.listen_addr.is_ipv4() == ip.is_ipv4()
+                        || self.listen_addr.is_ipv6() == ip.is_ipv6()
+                    {
+                        return Poll::Ready(TransportEvent::NewAddress {
+                            listener_id: self.listener_id,
+                            listen_addr: self.listen_multi_address(ip),
+                        });
+                    }
+                }
+                Ok(IfEvent::Down(inet)) => {
+                    let ip = inet.addr();
+                    if self.listen_addr.is_ipv4() == ip.is_ipv4()
+                        || self.listen_addr.is_ipv6() == ip.is_ipv6()
+                    {
+                        return Poll::Ready(TransportEvent::AddressExpired {
+                            listener_id: self.listener_id,
+                            listen_addr: self.listen_multi_address(ip),
+                        });
+                    }
+                }
+                Err(err) => {
+                    return Poll::Ready(TransportEvent::ListenerError {
+                        listener_id: self.listener_id,
+                        error: Error::Io(err),
+                    });
+                }
+            }
+        }
+
+        Poll::Pending
+    }
+
+    /// Constructs a [`Multiaddr`] for the given IP address that represents our listen address.
+    fn listen_multi_address(&self, ip: IpAddr) -> Multiaddr {
```

Comment:
```suggestion
    fn listen_multiaddress(&self, ip: IpAddr) -> Multiaddr {
```
Nit pick, I don't think we separate the two anywhere else.